pkg/internal/testmode/grpc-server-testmode.go

```diff
@@ -0,0 +1,214 @@
+	t.Setenv(grpcModeEnvVar, GrpcDockerMode)
+	for _, opt := range opts {
+		opt(t)
+	}
+}
+
+// Unwrap unwraps the mode to the expected type
+func Unwrap[M GrpcMode](mode Mode) M {
+	if mode == nil {
+		panic("mode to unwrap cannot be nil")
+	}
+	if m, ok := mode.(M); ok {
+		return m
+	}
+	panic("mode to unwrap is not of expected type")
+}
+
+// LocalGrpcMode is the mode to use a grpc server with typescript started locally.
+type LocalGrpcMode struct {
+	Host string
+	Port int
+}
+
+// ModeName returns the mode name
+func (m LocalGrpcMode) ModeName() string {
+	return GrpcLocalMode
+}
+
+func (m LocalGrpcMode) read() Mode {
+	modeName := os.Getenv(grpcModeEnvVar)
+	if modeName != m.ModeName() {
+		return nil
+	}
+
+	m.Host = os.Getenv(localGrpcModeHostEnvKey)
+	if m.Host == "" {
+		m.Host = "localhost"
+	}
+
+	m.Port = 50050
+	port := os.Getenv(localGrpcModePortEnvKey)
+	if port != "" {
+		var err error
+		m.Port, err = to.Int(port)
+		if err != nil {
+			panic(fmt.Errorf("invalid port number %s set on %s environment variable: %w", port, localGrpcModePortEnvKey, err))
+		}
+	}
+
+	return m
+}
+
+// DockerGrpcMode is the mode to start in docker a grpc server with typescript.
+type DockerGrpcMode struct {
+	Build     bool
+	Repo      string
+	Tag       string
+	KeepImage bool
+}
+
+func (m DockerGrpcMode) ModeName() string {
+	return GrpcDockerMode
+}
+
+func (m DockerGrpcMode) read() Mode {
+	modeName := os.Getenv(grpcModeEnvVar)
+	if modeName != "" && modeName != m.ModeName() {
+		return nil
+	}
+
+	buildStr := os.Getenv(dockerGrpcModeBuildImageEnvKey)
+	if buildStr != "" {
+		var err error
+		m.Build, err = to.BoolFromString(buildStr)
+		if err != nil {
+			panic(fmt.Errorf("invalid build value %s set on %s environment variable: %w", buildStr, dockerGrpcModeBuildImageEnvKey, err))
+		}
+	} else {
+		m.Build = true
+	}
+
+	m.Repo = os.Getenv(dockerGrpcModeImageRepoEnvKey)
+	if m.Repo == "" {
+		m.Repo = "go-middleware-regression-test-grpc-server"
+	}
+
+	m.Tag = os.Getenv(dockerGrpcModeImageTagEnvKey)
+	if m.Tag == "" {
+		m.Tag = "latest"
+	}
+
+	keepStr := os.Getenv(dockerGrpcModeKeepImageEnvKey)
+	if keepStr != "" {
+		var err error
+		m.KeepImage, err = to.BoolFromString(keepStr)
+		if err != nil {
+			panic(fmt.Errorf("invalid keep image value %s set on %s environment variable: %w", keepStr, dockerGrpcModeKeepImageEnvKey, err))
+		}
+	} else {
+		m.KeepImage = false
+	}
+	return m
+}
+
+// GetGrpcServerMode returns the test grpc server mode
+func GetGrpcServerMode() Mode {
+	modeReaders := seq.Of[modeReader](LocalGrpcMode{}, DockerGrpcMode{})
+
+	modes := seq.Map(modeReaders, func(mode modeReader) Mode {
+		return mode.read()
+	})
+
+	foundMode := seq.Find(modes, notNil)
+
+	mode, err := foundMode.OrErrorGet(func() error {
+		return fmt.Errorf("failed to determine grpc server mode based on env %s: %s", grpcModeEnvVar, os.Getenv(grpcModeEnvVar))
+	})
+	if err != nil {
+		panic(err)
+	}
+	return mode
+}
```

Comment:
I think, you overcomplicated this and unnecessarily used custom/common stuff and iterators.
```suggestion
func GetGrpcServerMode() Mode {
	mode := LocalGrpcMode{}.read()
	if mode != nil {
		return mode
	}
	
	mode = DockerGrpcMode{}.read()
	if mode != nil {
		return mode
	}

	panic(fmt.Errorf("failed to determine grpc server mode based on env %s: %s", grpcModeEnvVar, os.Getenv(grpcModeEnvVar)))
}
```